File: legacy/java/components/StrataFoundation/src/strata/foundation/injection/ProviderBasedBinder.java
// ##########################################################################
// # File Name:	ProviderBasedBinder.java
// #
// # Copyright:	2017, Sapientia Systems, LLC. All Rights Reserved.
// #
// # License:	This file is part of the StrataFoundation Framework.
// #
// #   			The StrataFoundation Framework is free software: you 
// #			can redistribute it and/or modify it under the terms of 
// #			the GNU Lesser General Public License as published by
// #    		the Free Software Foundation, either version 3 of the 
// #			License, or (at your option) any later version.
// #
// #    		The StrataFoundation Framework is distributed in the 
// #			hope that it will be useful, but WITHOUT ANY WARRANTY; 
// #			without even the implied warranty of MERCHANTABILITY or 
// #			FITNESS FOR A PARTICULAR PURPOSE.  See the GNU Lesser 
// #			General Public License for more details.
// #
// #    		You should have received a copy of the GNU Lesser 
// #			General Public License along with the StrataFoundation
// #			Framework. If not, see http://www.gnu.org/licenses/.
// ##########################################################################

package strata.foundation.injection;

import java.lang.annotation.Annotation;
import javax.inject.Provider;

/****************************************************************************
 * 
 * @author 		
 *     Sapientia Systems
 * @conventions	
 *     <a href="{@docRoot}/NamingConventions.html">Naming Conventions</a>
 */
public 
class ProviderBasedBinder<T>
    extends AbstractBinder<T>
{
    private final Provider<? extends T> itsProvider;
    
    /************************************************************************
     * Creates a new ProviderBasedBinder. 
     *
     * @param builder
     * @param provider
     * @param scope
     */
    public 
    ProviderBasedBinder(
        ISourceBindingBuilder<T>       builder,
        Provider<? extends T>       provider,
        Class<? extends Annotation> scope)
    {
        super( builder,scope );
        itsProvider = provider;
    }

    /************************************************************************
     * {@inheritDoc} 
     */
    @Override
    public void 
    bind()
    {
        getBuilder()
            .toProvider( itsProvider )
            .withScope( getScope() );
    }

}

// ##########################################################################

File: legacy/csharp/components/Strata.Foundation/Injection/ProviderBasedBinder.cs
//  ##########################################################################
//  # File Name: ProviderBasedBinder.cs
//  # Copyright: 2017, Sapientia Systems, LLC.
//  ##########################################################################

using System;
using SystemEx.Injection;

namespace Strata.Foundation.Injection
{
    //////////////////////////////////////////////////////////////////////////
    /// <summary>
    /// 
    /// </summary>
    ///  
    public
    class ProviderBasedBinder<T>:
        AbstractBinder<T>
    {
        protected IProvider<T> Provider { get; set; }

        //////////////////////////////////////////////////////////////////////
        /// <summary>
        /// </summary>
        /// 
        public
        ProviderBasedBinder(
            ISourceBindingBuilder<T> builder,
            IProvider<T>             provider,
            Type                     scope): 
            base(builder,scope)
        {
            Provider = provider;
        }

        //////////////////////////////////////////////////////////////////////
        /// <inheritDoc/>
        /// 
        public override void 
        Bind()
        {
            Builder
                .ToProvider(Provider)
                .WithScope(Scope);
        }
    }
}

//  ##########################################################################
